src/RoslynInsertionTool/RoslynInsertionTool/RoslynInsertionTool.VisualStudioTeamServices.cs

```diff
@@ -395,21 +397,45 @@ private static async Task<string> DownloadBuildArtifactsAsync(BuildHttpClient bu
             var archiveDownloadPath = Path.Combine(tempDirectory, artifact.Name);
             Console.WriteLine($"Downloading artifacts to {archiveDownloadPath}");
 
-            Stopwatch watch = Stopwatch.StartNew();
+            var artifactType = artifact.Resource.Type;
+            Debug.Assert(artifactType.Equals("container", StringComparison.OrdinalIgnoreCase) || artifactType.Equals("pipelineArtifact", StringComparison.OrdinalIgnoreCase));
 
-            using (Stream s = await buildClient.GetArtifactContentZipAsync(Options.ComponentBuildProjectNameOrFallback, build.Id, artifact.Name, cancellationToken))
-            using (var ms = new MemoryStream())
+            Stopwatch watch = Stopwatch.StartNew();
+            if (artifactType.Equals("container", StringComparison.OrdinalIgnoreCase))
             {
-                await s.CopyToAsync(ms);
-                using (ZipArchive archive = new ZipArchive(ms))
-                {
-                    archive.ExtractToDirectory(tempDirectory);
-                }
+                using var contentStream = await buildClient.GetArtifactContentZipAsync(Options.ComponentBuildProjectNameOrFallback, build.Id, artifact.Name, cancellationToken);
+                await ExtractToDirectoryAsync(contentStream, tempDirectory).ConfigureAwait(false);
+                Console.WriteLine($"Artifact download took {watch.ElapsedMilliseconds / 1000} seconds");
+                return archiveDownloadPath;
+            }
+            else
+            {
+                // When the published by Publish artifacts pipeline task, buildClient.GetArtifactContentZipAsync() is unable to get the content of the artifacts.
```

Comment:
Interesting that we download the artifact directly.  It would be nice if instead we could utilize the azdo pipeline download steps before running the tool, and giving the tool a path to the predownloaded artifact.

I don't know if that would break a bunch of things though.

Reply:
😄Probably the tool was originally created when Azp doesn't support yml? (Just too old)

Reply:
I only know we use it here https://dev.azure.com/dnceng/internal/_release?definitionId=102&view=mine&_a=releases but depends on the other teams's usage, if we move this to yml we might be pinged many times by other teams